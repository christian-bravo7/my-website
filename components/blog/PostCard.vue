<template>
  <div
    class="relative bg-white dark:bg-blue-700 bg-opacity-75 dark:bg-opacity-75 hover:bg hover:bg-opacity-100 dark-hover:bg-opacity-100 rounded-lg shadow-md transition-colors duration-200"
  >
    <template v-if="!placeholder">
      <div class="absolute inset-0 p-2 flex flex-col">
        <figure class="relative flex-1 overflow-hidden rounded-lg">
          <NuxtLink
            class="absolute inset-0 bg-cover bg-center post-image transition-transform duration-200"
            :style="{ backgroundImage: `url(${backgroundImage})` }"
            :to="path"
          />
        </figure>
        <div class="flex flex-col items-start py-2">
          <div class="w-full">
            <div
              v-if="tags.length"
              class="float-right ml-4"
            >
              <span
                v-for="(tag, index) in tags"
                :key="index"
                class="py-1 px-2 text-xs font-bold ml-2 select-none capitalize rounded-full bg-pink-800 text-gray-50"
              >
                {{ tag }}
              </span>
            </div>
            <NuxtLink
              class="inline text-xl font-bold hover:underline text-gray-900 dark:text-gray-200 select-none"
              :class="{ 'text-2xl': first }"
              :to="path"
            >
              {{ title }}
            </NuxtLink>
          </div>
          <p
            class="line-clamp text-sm mt-2 text-gray-900 dark:text-gray-200"
            :class="{ 'text-base': first }"
          >
            {{ description }}
          </p>
        </div>
      </div>
    </template>
    <template v-else>
      <div class="absolute inset-0 p-2 flex flex-col items-center justify-center select-none">
        <p class="text-gray-600 dark:text-gray-400 text-center">
          Writing cool things...
        </p>
        <p class="text-gray-600 dark:text-gray-400 text-center">
          Come back soon :)
        </p>
      </div>
    </template>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'nuxt-property-decorator';

@Component
export default class PostCard extends Vue {
  @Prop({ type: String })
  readonly backgroundImage!: string;

  @Prop({ type: String })
  readonly path!: string;

  @Prop({ type: String })
  readonly title!: string;

  @Prop({ type: String })
  readonly description!: string;

  @Prop({ type: Boolean })
  readonly first!: boolean;

  @Prop({ type: Boolean, default: false })
  readonly placeholder!: boolean;

  @Prop({ type: Array, default: () => [] })
  readonly tags!: Array<string>;
}
</script>

<style lang="scss" scoped>
.line-clamp {
  display: -webkit-box;
  overflow: hidden;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
}

.post-image {
  &:hover {
    transform: scale(1.2);
  }
}

html.dark-mode {
  .post-image {
    filter: brightness(0.9);
  }
}

</style>
