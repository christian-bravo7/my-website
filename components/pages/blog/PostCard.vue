<template>
  <div
    class="relative | bg-white dark:bg-dark-theme-900 bg-opacity-75 dark:bg-opacity-75 | hover:bg-opacity-100 dark-hover:bg-opacity-100 hover:shadow-xl | group | rounded-xl shadow-md | transition-all duration-300"
  >
    <div class="absolute inset-0 flex flex-col">
      <figure class="relative flex-1 overflow-hidden rounded-xl">
        <NuxtLink
          class="absolute inset-0 | bg-cover bg-center | group-hover:transform group-hover:scale-110 | transition-transform duration-300"
          :style="{ backgroundImage: `url(${image})` }"
          :to="path"
        />
      </figure>
      <div class="flex flex-col items-start p-4">
        <div class="flex justify-between items-center | w-full | mb-4">
          <div
            v-if="tags.length"
          >
            <AppPill
              v-for="(tag, index) in tags"
              :key="index"
              :text="tag"
            />
          </div>
          <div class="text-xs | text-gray-500">
            {{ formattedDate }}
          </div>
        </div>
        <NuxtLink
          class="inline-block text-2xl font-bold hover:underline text-gray-900 dark:text-gray-200 select-none"
          :class="{ 'text-3xl': first }"
          :to="path"
        >
          {{ title }}
        </NuxtLink>
        <p
          class="line-clamp text-sm mt-2 text-gray-900 dark:text-gray-200"
          :class="{ 'text-base': first }"
        >
          {{ description }}
        </p>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import dayjs from 'dayjs';
import { Component, Prop, Vue } from 'nuxt-property-decorator';

@Component
export default class PostCard extends Vue {
  @Prop({ type: String, required: true })
  readonly image!: string;

  @Prop({ type: String, required: true })
  readonly path!: string;

  @Prop({ type: String, required: true })
  readonly title!: string;

  @Prop({ type: String, required: true })
  readonly description!: string;

  @Prop({ type: Array, required: true })
  readonly date!: Array<number>;

  @Prop({ type: Boolean, default: false })
  readonly first!: boolean;

  @Prop({ type: Array, default: () => [] })
  readonly tags!: Array<string>;

  get formattedDate (): string {
    return dayjs(this.date as any).format('MMM DD, YYYY');
  }
}
</script>

<style lang="scss" scoped>
.line-clamp {
  display: flex;
  -webkit-line-clamp: 2;
  overflow: hidden;
}

</style>
