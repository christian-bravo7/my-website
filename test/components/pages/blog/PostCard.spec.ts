import { mount } from '@vue/test-utils';

import PostCard from '@/components/pages/blog/PostCard.vue';

describe('PostCard component', () => {
  it('should render correclty', () => {
    const wrapper = mount(PostCard, {
      propsData: {
        image: 'image.png',
        path: 'something-cool',
        title: 'awesome post',
        description: 'this is the amazing post ever',
        date: [2021, 2, 31],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
