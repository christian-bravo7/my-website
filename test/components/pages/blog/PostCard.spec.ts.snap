// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard component should render correclty 1`] = `
<div class="relative | bg-white dark:bg-dark-theme-900 bg-opacity-75 dark:bg-opacity-75 | hover:bg-opacity-100 dark-hover:bg-opacity-100 hover:shadow-xl | group | rounded-xl shadow-md | transition-all duration-300">
  <div class="absolute inset-0 flex flex-col">
    <figure class="relative flex-1 overflow-hidden rounded-xl"><a class="absolute inset-0 | bg-cover bg-center | group-hover:transform group-hover:scale-110 | transition-transform duration-300" style="background-image: url(image.png);"></a></figure>
    <div class="flex flex-col items-start p-4">
      <div class="flex justify-between items-center | w-full | mb-4">
        <!---->
        <div class="text-xs | text-gray-500">
          Mar 03, 2021
        </div>
      </div> <a class="inline-block text-2xl font-bold hover:underline text-gray-900 dark:text-gray-200 select-none">
        awesome post
      </a>
      <p class="line-clamp text-sm mt-2 text-gray-900 dark:text-gray-200">
        this is the amazing post ever
      </p>
    </div>
  </div>
</div>
`;
